### ISSUE TITLE
Lower INT format support

### ISSUE THREAD

**Issue by yaozhewei** [pytorch/pytorch#74627]
### 🚀 The feature, motivation and pitch

Currently, pytorch officially supports INT8 format. However, recently, the newer GPUs (like T4/A100) already supports INT4 tensorcore. Meanwhile, from machine learning community, binary/ternary/int4 algorithms are well developed. It is the time to support lower INT format to connect research to product by Pytorch. I collaborated with Huggingface and Huggingface team is also looking for this feature as well.

**Comment by stas00** [comment]
For more context please see [Mixture-of-Quantization: A novel quantization approach for reducing model size with minimal accuracy impact](https://www.deepspeed.ai/2021/05/04/MoQ.html).

So currently with only `int8` supported by pytorch, that's as low as we can go when quantizing the model. But if `int4` were to be available a further halving of the model size would be possible.

And thank you!

**Comment by Si13x** [comment]
Voicing my support for adding this feature to Pytorch -- in my opinion, at this point one of the biggest issues with large ML models is making them feasible to run on small-scale labs (and ideally, even home labs for the sufficiently enthusiastic and determined).  INT4 support looks like a very promising tool in achieving this goal of democratizing access, particularly as the install base of GPUs that can take advantage of it continues to grow.

**Comment by VitalyFedyunin** [comment]
Adding to Triage Review to raise awareness about this popular request

**Comment by EricSteinberger** [comment]
+1

**Comment by soumith** [comment]
@stas00 thanks, this has gotten more urgent over time -- and especially in the last 3 months.
@ezyang and @gchanan were also getting requests from quite a few people with the upcoming FP8 and FP4 standards. It gets more complex with FP8 because the number of bits for mantissa and exponent are variable too, another parameter.

The plan we settled on was to introduce `bit1, bit2, bit4, bit8, bit16`, etc. to first unblock everyone on the dtype needs.

@angelayi sent a first PR for `bit16` 12 days ago: https://github.com/pytorch/pytorch/pull/88594

More to come.

**Comment by ezyang** [comment]
Note that for sub byte bits, it will be easier for us to do bits1x8, bits2x4 and bits4x2, as it permits to keep individually addressable elements (since you can't use pointers to reference sub-byte locations)

**Comment by stas00** [comment]
@soumith, thank you very much for the follow up - that's an awesome lineup of bits you have! 

Looking forward to using those.

Also please note pytorch solutions like [Mixture-of-Quantization](https://www.deepspeed.ai/tutorials/MoQ-tutorial/) that can use any bit1..bit16, including all the intermediary bits. Surely there is no yet hardware support for say int12, so I'm not sure it needs to be natively supported in pytorch, but I thought I'd share this FYI, in case it affects your planning.

**Comment by ezyang** [comment]
> Also please note pytorch solutions like [Mixture-of-Quantization](https://www.deepspeed.ai/tutorials/MoQ-tutorial/) that can use any bit1..bit16, including all the intermediary bits

Power of 2 bits, or you mean literally every intermediary bit (like 7-bit, etc)?

> As you have to support FP8's 5/3 and 3/5 splits, perhaps FP12, could use the same format and simply define 6/6 or something similar - so FP8 could be a special case of a fpx format that defines its mantissa and exponent during the tensor creation.

Yeah, this is one of the reasons why we're opting for uninterpreted bits to start, it's still not entirely clear how the splits will be represented in UI, so we want to get some experience with a light weight implementation to find the right UX and then codify it in C++ later.

**Comment by stas00** [comment]
> > Also please note pytorch solutions like [Mixture-of-Quantization](https://www.deepspeed.ai/tutorials/MoQ-tutorial/) that can use any bit1..bit16, including all the intermediary bits
> 
> Power of 2 bits, or you mean literally every intermediary bit (like 7-bit, etc)?

Yes, any discrete number of bits between 1 and 16 can be used, e.g. see the table in:
https://www.deepspeed.ai/tutorials/MoQ-tutorial/#finetuning-results

But I am not sure this needs to be exposed on the pytorch side - as this is all done directly in CUDA C++ from what I understand.

**Comment by jerryzh168** [comment]
how are we using these new types? wondering if we should first list all the things we want to do with these dtypes before adding this to core. In quantization we would like to use existing dtypes (like uint8, int8) in the future instead of adding new quantized dtypes.

I can think of at least two different scenarios of usage (taking int4 as an example):
1. just represent some tensor/op is int4, but the real operators are implemented in somewhere else (e.g. TensorRT library)
2. we want to implement pytorch ops in int4 in pytorch native_functions.yaml

I feel we only need new dtypes in PyTorch core if we are targeting 2., is this the current target?

**Comment by stas00** [comment]
That's not necessarily true, I think even if pytorch doesn't support say `int4` kernels, 2 3rd party systems might need to communicate tensors to each other, so the very basic need is to at least have `torch.init4` as a dtype. so that a user could create this tensor and then pass it to a system that has int4-based kernels.

For example a 3rd party library may return an int4-tensor and the user should be able to say write/load this tensor to/from disk.

So to refer to your list, I'd say (1) is needed on its own and is probably a low-hanging fruit. (2) requires actual specific requests to what kernels should be ported to support int4-ops.

does it make sense?

**Comment by jerryzh168** [comment]
> That's not necessarily true, I think even if pytorch doesn't support say `int4` kernels, 2 3rd party systems might need to communicate tensors to each other, so the very basic need is to at least have `torch.init4` as a dtype. so that a user could create this tensor and then pass it to a system that has int4-based kernels.
> 
> For example a 3rd party library may return an int4-tensor and the user should be able to say write/load this tensor to/from disk.
> 
> So to refer to your list, I'd say (1) is needed on its own and is probably a low-hanging fruit. (2) requires actual specific requests to what kernels should be ported to support int4-ops.
> 
> does it make sense?

Yeah the use case makes sense, but I feel that doesn't really require us to add new dtypes in PyTorch though, since the meaning of the Tensor can be encoded in the operator that's consuming the Tensor as well, to give a concrete example:

```
Tensor int4_add(Tensor a, Tensor b)
```
We can have `a` and `b` to be stored as `torch.uint8` Tensor, but when we pass them to `int4_add` operator, that's implemented in a third party library, the third party library can re-interpret Tensor `a` and `b` as int4 Tensors, even though they are stored as `torch.uint8`. In this case the "meaning" of the Tensor is in the operator itself (since we are calling this Tensor with int4_add)

Also I would suspect that even if we are implementing sub-byte dtypes in PyTorch, they will be backed up by `torch.uint8`/`toch.int8` anyways?

btw, it doesn't prevent us from adding python-only dtypes, but I'm still not too clear about the necessity of adding this in PyTorch core and treating these the same way as other dtypes like `torch.int8` while the only thing we need is just to represent them.

**Comment by soumith** [comment]
@jerryzh168 it's useful to have `bit4x2` tensors precisely to express intent way better than overloading a `torch.uint8`. the `bit` datatype is saying these are bits and everything else is off bets. Separately, we'll have to talk about storing metadata in a `bit` Tensor because how would you store exponent scaling information for Float8 on a tensor otherwise

**Comment by ezyang** [comment]
@stas00 huh. So let's say I'm doing 5bit. They will pack it, so in 40 bits (5 bytes) I will store 8 5-bit quantities, is that right?

**Comment by stas00** [comment]
@ezyang, I'm not familiar with the internals. Let's ask @RezaYazdaniAminabadi who is one of the main developers of MoQ. 

Reza, could you please answer Edward's question?

For context this feature request has grown from asking for adding just `torch.int4` dtype to what other dtypes need to be supported and I brought up MoQ where any number of bits might be used. And while we all agree that the lengths of bits that aren't supported in the hardware (int3) shouldn't have a dedicated dtype, the discussion is how pytorch should support those if at all. FP8 already pushed the boundaries with its shifting sub-types and a need for a scaling factor, so I think the developers are trying to gather the information about possible needs before deciding on how to implement new dtypes. Thank you.

**Comment by jerryzh168** [comment]
> @jerryzh168 it's useful to have `bit4x2` tensors precisely to express intent way better than overloading a `torch.uint8`. the `bit` datatype is saying these are bits and everything else is off bets. Separately, we'll have to talk about storing metadata in a `bit` Tensor because how would you store exponent scaling information for Float8 on a tensor otherwise

I feel it makes sense to add `bits4x2` etc. my question is on the implementation side. Just want to make sure we have considered the alternatives.

Alternative: Define the new dtypes in python (e.g. bits4x2), that can be backed up by existing dtypes (e.g. uint8)
Pros
* adding new dtypes is easy (no change to pytorch core (as seen in https://github.com/pytorch/pytorch/pull/88594))
* the backends that integrate with PyTorch (e.g. executorch) don't need to know about these new dtypes
Cons
* not clear from dtype of tensor itself what the tensor means in C++ land, but the meaning will be clear from the operator that is using the Tensor. e.g. `add_float8(a, a_scaling_factor, b, b_scaling_factor)`, even though `a` and `b` are `torch.uint8` we still know it's supposed to represent `float8` tensor, similarly, if we have `add_4bit(a, a_metadata, b, b_metadata)` we will know `a` and `b` are 4 bit Tensors

**Comment by ezyang** [comment]
For bits4x2 and other sub-byte size types, there's a choice about how you want construction/indexing/queries to work.

Let's say I ask for `torch.empty((4, 2), dtype=bits4x2)`. Does this mean I have eight bits4 entries, or do I have eight bits4x2 entries? The decision that QUInt4x2/QUInt2x4 types made is that you have eight bits4 entries; and furthermore we will pad the inner dimension:

```
int64_t get_sub_byte_tensor_size(IntArrayRef sizes, size_t dtype_itemsize, at::ScalarType t) {
  // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
  int64_t element_per_byte;
  switch(t) {
    case at::ScalarType::QUInt4x2:
      element_per_byte = 2;
      break;
    case at::ScalarType::QUInt2x4:
      element_per_byte = 4;
      break;
    default:
      element_per_byte = 1;
  }   
  // zero dim tensor
  if (sizes.size() == 0) {
    return c10::multiply_integers(sizes) * dtype_itemsize;
  }
  // Consider most inner dim as cols
  int64_t cols = sizes.at(sizes.size()-1);
  int64_t bytes_per_row = cols * dtype_itemsize;
  // align qtensor most inner dim, compute ceil (bytes_per_row / element_per_byte)
  return c10::multiply_integers(IntArrayRef(sizes.data(), sizes.size() - 1)) * at::ceil_div(bytes_per_row, element_per_byte);
}
```

Was this extra logic a mistake? If PyTorch doesn't need to handle the xN data types specially, I suppose there is not much point in having them, you can just have the byte-size types only. But it seems to me that the request was always for PyTorch to take care of the grungy bookkeeping necessary for sub bytes computation.

For bits8 versus uint8, it is mostly a matter of communicating intent. With bits8 you can ensure addition, e.g., is not defined, because it might be something weird, whereas with uint8 you may be tempted to add two tensors together (with disastrous results).

**Comment by AutomaticHourglass** [comment]
As a mac user, I would love to have this functionality on `mps` platform and totally supporting this. As I have developed AVX2 and AVX512 in the past, I can also help with the transition.

**Comment by vadimkantorov** [comment]
bits data types are finally in, and need for int4 quantized types support for CPU and CUDA is even higher now :)

**Comment by stas00** [comment]
update: I have resurfaced this old Issue on twitter and Mark pointed to where this issue is being addressed: https://github.com/pytorch-labs/ao/blob/aa9463960ad22c056b5e7e7472ad456e23756a9f/torchao/dtypes/uint4.py

very cool!

@yaozhewei, is this something you could use?

**Comment by stas00** [comment]
Looking at FasterTransformer - I see a lot of uint4 there:
https://github.com/search?q=repo%3ANVIDIA%2FFasterTransformer%20int4&type=code

it looks like they pack it into int32:
https://github.com/NVIDIA/FasterTransformer/blob/df4a7534860137e060e18d2ebf019906120ea204/src/fastertransformer/kernels/custom_ar_kernels.cu#L68

**Comment by Xynonners** [comment]
I'd like to also know if there's any progress being made on this, since smaller dtypes would be great for my current usecase (low-range floats)

**Comment by ezyang** [comment]
cc @jerryzh168

**Comment by vadimkantorov** [comment]
Interestingly, ONNX added some representations of INT4/UINT4 in https://github.com/pytorch/pytorch/issues/121258

**Comment by jerryzh168** [comment]
@stas00, @Xynonners and @vadimkantorov: this is added as a prototype that you can use to build any new tensors using uint4 dtype: https://github.com/pytorch-labs/ao/blob/main/torchao/dtypes/uint4.py, but the support is not complete yet, and feel free to add new aten ops, in the future our plan is to make this more complete and move this to pytorch core

**Comment by vadimkantorov** [comment]
@ezyang UINT4 btw might also trigger the same questions as wide/tiled dtypes :)

**Comment by msaroufim** [comment]
For people who made it here would recommend following discussion on https://github.com/pytorch-labs/ao/issues/47

**Comment by ionutmodo** [comment]
I am following up on this to support implementing this feature because it would be super useful to design low memory optimizers.

**Comment by msaroufim** [comment]
@ionutmodo we managed to implement low bit optimizers using pure pytorch here which includes int8/4 and fp8 https://github.com/pytorch/ao/tree/main/torchao/prototype/low_bit_optim
